### ISSUE TITLE
Allow TFDS datasets defined outside TFDS

### ISSUE THREAD

**Issue by pierrot0** [tensorflow/datasets#704]
A DatasetBuilder class can be defined outside TFDS (ie: not committed to tfds repository).
There might however be a few tricks, and those are not documented.

We should document the process of writing such dataset, and provide the necessary hooks.
For example, download_and_prepare needs checksums, which shouldn't have to be registered in TFDS (see issue #580). For that particular case, we could write a method to add additional checksum directories, which each dataset can call before download_and_prepare is.

There might be other small changes needed. Those should be clarified when writing the documentation.

**Comment by us** [comment]
I'm starting to work on it!

**Comment by craymichael** [comment]
Yes to this! Currently, I've used the following nasty lines to mitigate this issue

```python
from tensorflow_datasets.core.download import checksums
checksums._CHECKSUM_DIRS.append(os.path.join(os.path.dirname(__file__),
                                             'url_checksums'))
```

Getting to testing to work properly (e.g. `fake_examples` data) was equally not great.

**Comment by Conchylicultor** [comment]
@craymichael for the fake data, you can set-up your own custom dir using `EXAMPLE_DIR`. It didn't worked for you ?

```
class MyDatasetTest(DatasetBuilderTestCase):
  EXAMPLE_DIR = 'path/to/fakedata'
```

@us Thanks for adding this. if you have time, could you create a new section [after this]( https://www.tensorflow.org/datasets/add_dataset#adding_the_dataset_to_tensorflowdatasets) `Adding the dataset outside tensorflow/datasets` detailing the steps required to define a dataset outside `tfds` ? Using `tfds.download.add_checksums_dir` and `TESTING_DIR` for tests.

**Comment by craymichael** [comment]
@Conchylicultor Yes, that is what I ended up doing, but for some reason I thought the test utilities would create the directory/data for me as a subset of the original data based on the dummy parameters I utilized. I think at the time of writing I was more confused as to why `DL_EXTRACT_RESULT` was necessary

**Comment by olk** [comment]
The documentation is incomplete!
You have to document the naming format of the file that holds the checksums (== name of the dataset class in lower cases with under scores) and pass DownloadConfig(register_checksums=True) to
DownloadManager.download_and_prepare()

**Comment by Conchylicultor** [comment]
@olk sorry about that, please don't hesitate to send a PR with the additional info
